Fixed order of generated edges and guidance turns

oxidase opened the pull request:
# Issue

This step is required before implementation of #4674

The order of edges is fixed by the order of node-based graph nodes and adjacent intersection edges `incoming_edges` and `outgoing_edges`  and the implicit intersection connectivity matrix that is specified by a free function `isTurnAllowed`.

## Tasklist
 - [x] Re-enable via-way restrictions
 - [x] Check changes of generated turns for BER
 - [x]  Check changes of generated turns for CA
 - [x]  Check changes of generated turns for DE
 - [x] Cleanup the PR
 - [x] Squash commits of the PR
 - [x] add regression / cucumber cases (see docs/testing.md)
 - [x] Add CHANGELOG entry
 - [x] review
 - [x] adjust for comments

## Requirements / Relations
Blocks #4674

oxidase commented:
Compared generated turn instruction with respect to master (modifications are additions or deletions only):

* [BER](https://github.com/Project-OSRM/osrm-backend/files/1518151/berlin-diff.txt) 8 modifications in 401144 generated instructions

*  [CA](https://github.com/Project-OSRM/osrm-backend/files/1518153/california-diff.txt) 186 modifications for 12151513 generated instructions

* [DE](https://github.com/Project-OSRM/osrm-backend/files/1518147/germany-diff.txt) 583 modifications for 
30656676  generated instructions

The number of modified instructions is vanishingly small ~0.002% and almost all locations are either suppressed or invalid turns.

There are some manually selected locations that show  edge cases in bearings merging and must be fixed later:
* http://map.project-osrm.org/?z=18&center=52.547721%2C13.416141&loc=52.547966%2C13.416983&loc=52.547777%2C13.416898&hl=en&alt=0
![screenshot from 2017-11-30 15-41-26](https://user-images.githubusercontent.com/4421046/33436324-fe0d556e-d5e4-11e7-8f3c-1c19eaf5f7f0.png)

* http://map.project-osrm.org/?z=18&center=37.321675%2C-122.078914&loc=37.322039%2C-122.076457&loc=37.321788%2C-122.076210&hl=en&alt=0
![screenshot from 2017-11-30 15-42-14](https://user-images.githubusercontent.com/4421046/33436367-15eb4dc6-d5e5-11e7-9d82-200c9b35800e.png)

* http://map.project-osrm.org/?z=18&center=34.380286%2C-118.568627&loc=34.379868%2C-118.569029&loc=34.380483%2C-118.568519&hl=en&alt=0
![screenshot from 2017-11-30 15-43-01](https://user-images.githubusercontent.com/4421046/33436404-306b80c6-d5e5-11e7-93f0-58a21d23163a.png)


* http://map.project-osrm.org/?z=18&center=34.004865%2C-118.176354&loc=34.005223%2C-118.174401&loc=34.005405%2C-118.174042&hl=en&alt=0
![screenshot from 2017-11-30 15-43-36](https://user-images.githubusercontent.com/4421046/33436420-468f1610-d5e5-11e7-8bcd-ac38ec3c13f3.png)


* http://map.project-osrm.org/?z=18&center=34.005258%2C-118.174015&loc=34.005467%2C-118.174133&loc=34.005032%2C-118.173827&hl=en&alt=0
![screenshot from 2017-11-30 15-44-31](https://user-images.githubusercontent.com/4421046/33436462-6d8d6adc-d5e5-11e7-8aa7-de0681e3d63a.png)


* http://map.project-osrm.org/?z=18&center=47.928820%2C11.503587&loc=47.928115%2C11.503635&loc=47.929643%2C11.502562&hl=en&alt=0
![screenshot from 2017-11-30 15-45-24](https://user-images.githubusercontent.com/4421046/33436499-83f7f242-d5e5-11e7-8dd3-312e59d380fa.png)


* http://map.project-osrm.org/?z=18&center=48.230855%2C11.669717&loc=48.229418%2C11.668124&loc=48.229661%2C11.667958&hl=en&alt=0
![screenshot from 2017-11-30 15-46-03](https://user-images.githubusercontent.com/4421046/33436524-9be37f0c-d5e5-11e7-8d70-25c1020cf1db.png)

* http://map.project-osrm.org/?z=18&center=51.675110%2C12.273150&loc=51.674075%2C12.272034&loc=51.674401%2C12.271836&hl=en&alt=0
![screenshot from 2017-11-30 15-46-34](https://user-images.githubusercontent.com/4421046/33436542-acfea848-d5e5-11e7-8d05-102f469c3bb3.png)

TheMarex commented on include/extractor/guidance/statistics_handler.hpp:
```diff
@@ -84,21 +84,23 @@ class StatisticsHandler final : public IntersectionHandler
     mutable std::mutex lock;
-    mutable std::unordered_map<TurnType::Enum, std::uint64_t> type_hist;
-    mutable std::unordered_map<DirectionModifier::Enum, std::uint64_t> modifier_hist;
+    mutable std::map<TurnType::Enum, std::uint64_t> type_hist;
```
Is there a particular reason we need to preserve the insertion order here? Just curious.

oxidase replied:
yes, `map` preserves order, so it is easier to do line-by-line comparison with `diff`

TheMarex commented on src/extractor/intersection/intersection_analysis.cpp:
```diff
@@ -0,0 +1,721 @@
+    }
+
+    // Enforce ordering of outgoing edges
+    std::sort(result.begin(), result.end());
```
This sequence should be sorted by construction.

oxidase replied:
it depends on `GetAdjacentEdgeRange` implementation, but can add `std::is_sorted` assertion in both places

danpat commented:
@oxidase minor - can you add a CHANGELOG entry for this?  It'll help when reviewing changes for the next release.

TheMarex commented on src/extractor/intersection/intersection_analysis.cpp:
```diff
@@ -0,0 +1,808 @@
+
+    for (const auto outgoing_edge : graph.GetAdjacentEdgeRange(intersection_node))
+    {
+        // TODO: to use TurnAnalysis all outgoing edges are required, to be uncommented later
```
I don't quite understand yet how we could skip `reversed` edges in the future. How will this be possible?

TheMarex commented on src/extractor/intersection/intersection_analysis.cpp:
```diff
@@ -0,0 +1,808 @@
+            return true;
+
+        // Allow U-turn if the incoming edge has a U-turn lane
+        const auto &incoming_edge_annotation_id = graph.GetEdgeData(from.edge).annotation_data;
```
We should probably discuss if using the presence of a `uturn` lane is a good idea here, or if we should base this on country specific settings. Making the accessibility depend on turn lanes is just quite ugly.

In general, we should investigate how often this code path is even triggered - I have a feeling that most streets with a u-turn lane also have segregated ways for both directions (so it is not a "simple" uturn).

oxidase replied:
added a TODO line here, but i would keep it as it is now

TheMarex commented on unit_tests/extractor/intersection_analysis_tests.cpp:
```diff
@@ -0,0 +1,315 @@
+using InputEdge = util::NodeBasedDynamicGraph::InputEdge;
+using Graph = util::NodeBasedDynamicGraph;
+
+BOOST_AUTO_TEST_CASE(simple_intersection_connectivity)
```
Nice! Good to see some unit tests for this code.

oxidase commented on include/extractor/guidance/turn_lane_types.hpp:
```diff
@@ -111,8 +113,7 @@ transformTurnLaneMapIntoArrays(const LaneDescriptionMap &turn_lane_map)
     // adjacency array like structure to store the turn lane masks.
-    std::vector<std::uint32_t> turn_lane_offsets(turn_lane_map.data.size() +
-                                                 2); // empty ID + sentinel
+    std::vector<std::uint32_t> turn_lane_offsets(turn_lane_map.data.size() + 1); // + sentinel
```
probably it is for `empty ID` but an empty entry is already included into `turn_lane_map.data`

oxidase commented on src/extractor/edge_based_graph_factory.cpp:
```diff
@@ -419,22 +421,43 @@ void EdgeBasedGraphFactory::GenerateEdgeExpandedEdges(
 
     TurnDataExternalContainer turn_data_container;
 
+    // TODO: add a MergableRoadDetector instance, to be deleted later
```
it is better to keep `MergableRoadDetector` for now, but it also influences intersections connectivity via edges merging and creating U-turns

oxidase commented on src/extractor/intersection/intersection_analysis.cpp:
```diff
@@ -0,0 +1,808 @@
+        road.edge = graph.GetTarget(next_edge) == road.node ? next_edge + 1 : next_edge;
+        road.node = next_node;
+        next_node = graph.GetTarget(road.edge);
+    }
```
This check always returns true because of an assignment  `query_edge = next_edge;` one line above and it was a comparison  of two identical data structures.

oxidase commented on src/extractor/guidance/mergable_road_detector.cpp:
```diff
@@ -165,8 +170,9 @@ bool MergableRoadDetector::EdgeDataSupportsMerge(
 bool MergableRoadDetector::IsTrafficLoop(const NodeID intersection_node,
```
yes, it is needed to prevent merges of loops like https://www.openstreetmap.org/?mlat=43.6174&mlon=7.05839#map=19/43.61695/7.05820

oxidase commented:
Checked modified instruction in SF https://gist.github.com/oxidase/f0492bd6ffff80a7f24a91a2f07a4602 
There are 45 modified instructions in 11 unique places and most changes are new U-turn instructions due to 701ac2c that also fixes #4704.

Also ce17c78144f8cd4bede87e996449ad65d149fe95 changes  `distance_to_extract` from 150 to 120 meters. This makes `are_parallel` check less dependent on roads divergence, like `South Van Ness Avenue` at https://www.openstreetmap.org/node/65299217

TheMarex commented:
@oxidase heads up I rebased this on master for a pre-release.

TheMarex commented:
I ran this PR on North America and Europe and found two locations that trigger an assertion:

Europe:
http://www.openstreetmap.org/?mlat=43.4399&mlon=5.3615#map=19/43.4399/5.3615

North-America:
http://www.openstreetmap.org/?mlat=38.9844&mlon=-104.763#map=19/38.9844/-104.763

Both locations trigger the same assertion in [`sliproad_handler.cpp:604`](https://github.com/Project-OSRM/osrm-backend/blob/d1022043ba4921bb009038fc1309d4b5d497e83a/src/extractor/guidance/sliproad_handler.cpp#L604). It is not obvious to me why this location should throw an assertion so I'll dig a little.

oxidase commented:
@TheMarex I have added a test case for the failing assertion and removed the assertion line in 8ab96541188e957375c9e007aa27ed496ddeaeab of #4756 that fixes a small regression in the last patch release 5.14.2

daniel-j-h commented:
cc @karenzshea ^ that's the branch triggering you tried to find a real-world use case for; maybe you have more context.

oxidase commented:
@TheMarex @daniel-j-h 5.14.3 release fixes assertions

oxidase commented:
The commit 154ceb1de99aa11371dc3ca45f1f9fa7dcaa9bee changes turn instructions at places on the diff map http://geojson.io/#id=gist:oxidase/221b0b27779fbb5ca662a35d699708bc&map=12/37.8198/-122.3956
